Crash in [@ org.mozilla.geckoview.GeckoResult$UncaughtException: at org.mozilla.geckoview.GeckoResult.dispatchLocked(GeckoResult.java)]

https://crash-stats.mozilla.org/signature/?product=Fenix&signature=org.mozilla.geckoview.GeckoResult%24UncaughtException%3A%20at%20org.mozilla.geckoview.GeckoResult.dispatchLocked%28GeckoResult.java%29

bz-1644999
```
org.mozilla.geckoview.GeckoResult$UncaughtException: java.lang.IllegalStateException: Can not perform this action after onSaveInstanceState
	at org.mozilla.geckoview.GeckoResult.dispatchLocked(GeckoResult.java:4)
	at org.mozilla.geckoview.GeckoResult.completeExceptionally(GeckoResult.java:4)
	at org.mozilla.geckoview.GeckoResult.lambda$thenInternal$2$GeckoResult(GeckoResult.java:10)
	at org.mozilla.geckoview.-$$Lambda$GeckoResult$gwCgOUK_EYQn2g6GolfZvo6A_WE.run(lambda)
	at org.mozilla.geckoview.GeckoResult.lambda$dispatchLocked$3(GeckoResult.java:2)
	at org.mozilla.geckoview.-$$Lambda$GeckoResult$PehjG2jgyDa_p37vZrKrSi2I94s.run(lambda)
	at android.os.Handler.handleCallback(Handler.java:815)
	at android.os.Handler.dispatchMessage(Handler.java:104)
	at android.os.Looper.loop(Looper.java:207)
	at android.app.ActivityThread.main(ActivityThread.java:5740)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:905)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:766)
Caused by: java.lang.IllegalStateException: Can not perform this action after onSaveInstanceState
	at androidx.fragment.app.FragmentManager.enqueueAction(FragmentManager.java:6)
	at androidx.fragment.app.BackStackRecord.commitInternal(BackStackRecord.java:14)
	at androidx.fragment.app.BackStackRecord.commit(BackStackRecord.java:1)
	at androidx.fragment.app.DialogFragment.show(DialogFragment.java:5)
	at -$$LambdaGroup$ks$UtaWBn0vzFfZ64mP0sa7JCFLt3Q.invoke(com.android.tools.r8.jetbrains.kotlin-style lambda group:83)
	at -$$LambdaGroup$ks$_HIPVZOvbmOxwps1M61uurZH_lk.invoke(com.android.tools.r8.jetbrains.kotlin-style lambda group:15)
	at mozilla.components.browser.engine.gecko.GeckoEngine$installWebExtension$$inlined$apply$lambda$3.onValue(GeckoEngine.kt:14)
	at org.mozilla.geckoview.GeckoResult.lambda$thenInternal$2$GeckoResult(GeckoResult.java:2)
```

--- a/components/feature/addons/src/main/java/mozilla/components/feature/addons/ui/AddonInstallationDialogFragment.kt
+++ b/components/feature/addons/src/main/java/mozilla/components/feature/addons/ui/AddonInstallationDialogFragment.kt
@@ -6,6 +6,7 @@ package mozilla.components.feature.addons.ui
 
 import android.annotation.SuppressLint
 import android.app.Dialog
+import android.graphics.Bitmap
 import android.graphics.Color
 import android.graphics.drawable.BitmapDrawable
 import android.graphics.drawable.ColorDrawable
@@ -25,6 +26,7 @@ import androidx.annotation.VisibleForTesting
 import androidx.appcompat.app.AppCompatDialogFragment
 import androidx.appcompat.widget.AppCompatCheckBox
 import androidx.core.content.ContextCompat
+import androidx.fragment.app.FragmentManager
 import kotlinx.android.synthetic.main.mozac_feature_addons_fragment_dialog_addon_installed.view.*
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
@@ -38,25 +40,32 @@ import mozilla.components.support.ktx.android.content.appName
 import mozilla.components.support.ktx.android.content.res.resolveAttribute
 import java.io.IOException
 
+@VisibleForTesting internal const val KEY_INSTALLED_ADDON = "KEY_ADDON"
 private const val KEY_DIALOG_GRAVITY = "KEY_DIALOG_GRAVITY"
 private const val KEY_DIALOG_WIDTH_MATCH_PARENT = "KEY_DIALOG_WIDTH_MATCH_PARENT"
 private const val KEY_CONFIRM_BUTTON_BACKGROUND_COLOR = "KEY_CONFIRM_BUTTON_BACKGROUND_COLOR"
 private const val KEY_CONFIRM_BUTTON_TEXT_COLOR = "KEY_CONFIRM_BUTTON_TEXT_COLOR"
 private const val KEY_CONFIRM_BUTTON_RADIUS = "KEY_CONFIRM_BUTTON_RADIUS"
+@VisibleForTesting internal const val KEY_ICON = "KEY_ICON"
+
 private const val DEFAULT_VALUE = Int.MAX_VALUE
 
 /**
  * A dialog that shows [Addon] installation confirmation.
  */
-class AddonInstallationDialogFragment(
-    private val addonCollectionProvider: AddonCollectionProvider
-) : AppCompatDialogFragment() {
+class AddonInstallationDialogFragment : AppCompatDialogFragment() {
     private val scope = CoroutineScope(Dispatchers.IO)
+    @VisibleForTesting internal var iconJob: Job? = null
     private val logger = Logger("AddonInstallationDialogFragment")
     /**
      * A lambda called when the confirm button is clicked.
      */
     var onConfirmButtonClicked: ((Addon, Boolean) -> Unit)? = null
+
+    /**
+     * Reference to the application's [AddonCollectionProvider] to fetch add-on icons.
+     */
+    var addonCollectionProvider: AddonCollectionProvider? = null
 
     private val safeArguments get() = requireNotNull(arguments)
 
@@ -90,6 +99,11 @@ class AddonInstallationDialogFragment(
                 KEY_CONFIRM_BUTTON_TEXT_COLOR,
                 DEFAULT_VALUE
             )
+
+    override fun onStop() {
+        super.onStop()
+        iconJob?.cancel()
+    }
 
     override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
         val sheetDialog = Dialog(requireContext())
@@ -144,7 +158,12 @@ class AddonInstallationDialogFragment(
                 requireContext().appName
             )
 
-        fetchIcon(addon, rootView.icon)
+        val icon = safeArguments.getParcelable<Bitmap>(KEY_ICON)
+        if (icon != null) {
+            rootView.icon.setImageDrawable(BitmapDrawable(resources, icon))
+        } else {
+            iconJob = fetchIcon(addon, rootView.icon)
+        }
 
         val allowedInPrivateBrowsing = rootView.findViewById<AppCompatCheckBox>(R.id.allow_in_private_browsing)
         allowedInPrivateBrowsing.setOnCheckedChangeListener { _, isChecked ->
@@ -188,9 +207,10 @@ class AddonInstallationDialogFragment(
     internal fun fetchIcon(addon: Addon, iconView: ImageView, scope: CoroutineScope = this.scope): Job {
         return scope.launch {
             try {
-                val iconBitmap = addonCollectionProvider.getAddonIconBitmap(addon)
+                val iconBitmap = addonCollectionProvider?.getAddonIconBitmap(addon)
                 iconBitmap?.let {
                     scope.launch(Dispatchers.Main) {
+                        safeArguments.putParcelable(KEY_ICON, it)
                         iconView.setImageDrawable(BitmapDrawable(iconView.resources, it))
                     }
                 }
@@ -204,6 +224,19 @@ class AddonInstallationDialogFragment(
                 logger.error("Attempt to fetch the ${addon.id} icon failed", e)
             }
         }
+    }
+
+    override fun show(manager: FragmentManager, tag: String?) {
+        // This dialog is shown as a result of an async operation (installing
+        // an add-on). Once installation succeeds, the activity may already be
+        // in the process of being destroyed. Since the dialog doesn't have any
+        // state we need to keep, and since it's also fine to not display the
+        // dialog at all in case the user navigates away, we can simply use
+        // commitAllowingStateLoss here to prevent crashing on commit:
+        // https://github.com/mozilla-mobile/android-components/issues/7782
+        val ft = manager.beginTransaction()
+        ft.add(this, tag)
+        ft.commitAllowingStateLoss()
     }
 
     @Suppress("LongParameterList")
@@ -224,11 +257,11 @@ class AddonInstallationDialogFragment(
             onConfirmButtonClicked: ((Addon, Boolean) -> Unit)? = null
         ): AddonInstallationDialogFragment {
 
-            val fragment = AddonInstallationDialogFragment(addonCollectionProvider)
+            val fragment = AddonInstallationDialogFragment()
             val arguments = fragment.arguments ?: Bundle()
 
             arguments.apply {
-                putParcelable(KEY_ADDON, addon)
+                putParcelable(KEY_INSTALLED_ADDON, addon)
 
                 promptsStyling?.gravity?.apply {
                     putInt(KEY_DIALOG_GRAVITY, this)
@@ -246,6 +279,7 @@ class AddonInstallationDialogFragment(
             }
             fragment.onConfirmButtonClicked = onConfirmButtonClicked
             fragment.arguments = arguments
+            fragment.addonCollectionProvider = addonCollectionProvider
             return fragment
         }
     }

--- a/components/feature/addons/src/test/java/mozilla/components/feature/addons/ui/AddonInstallationDialogFragmentTest.kt
+++ b/components/feature/addons/src/test/java/mozilla/components/feature/addons/ui/AddonInstallationDialogFragmentTest.kt
@@ -14,6 +14,7 @@ import androidx.appcompat.widget.AppCompatCheckBox
 import androidx.fragment.app.FragmentManager
 import androidx.fragment.app.FragmentTransaction
 import androidx.test.ext.junit.runners.AndroidJUnit4
+import kotlinx.coroutines.Job
 import kotlinx.coroutines.runBlocking
 import kotlinx.coroutines.test.TestCoroutineDispatcher
 import kotlinx.coroutines.test.TestCoroutineScope
@@ -21,18 +22,24 @@ import mozilla.components.feature.addons.Addon
 import mozilla.components.feature.addons.R
 import mozilla.components.feature.addons.amo.AddonCollectionProvider
 import mozilla.components.feature.addons.ui.AddonInstallationDialogFragment
+import mozilla.components.feature.addons.ui.KEY_ICON
+import mozilla.components.feature.addons.ui.KEY_INSTALLED_ADDON
 import mozilla.components.feature.addons.ui.translatedName
 import mozilla.components.support.test.mock
 import mozilla.components.support.test.robolectric.testContext
 import mozilla.components.support.test.rule.MainCoroutineRule
 import mozilla.components.support.test.whenever
 import org.junit.Assert.assertFalse
+import org.junit.Assert.assertNotNull
+import org.junit.Assert.assertNull
+import org.junit.Assert.assertSame
 import org.junit.Assert.assertTrue
 import org.junit.Assert.fail
 import org.junit.Rule
 import org.junit.Test
 import org.junit.runner.RunWith
 import org.mockito.Mockito
+import org.mockito.Mockito.`when`
 import org.mockito.Mockito.doReturn
 import org.mockito.Mockito.spy
 import org.mockito.Mockito.verify
@@ -51,7 +58,10 @@ class AddonInstallationDialogFragmentTest {
             "id", translatableName = mapOf(Addon.DEFAULT_LOCALE to "my_addon"),
             permissions = listOf("privacy", "<all_urls>", "tabs")
         )
-        val fragment = createAddonInstallationDialogFragment(addon, mock())
+        val mockedCollectionProvider = mock<AddonCollectionProvider>()
+        val fragment = createAddonInstallationDialogFragment(addon, mockedCollectionProvider)
+        assertSame(mockedCollectionProvider, fragment.addonCollectionProvider)
+        assertSame(addon, fragment.arguments?.getParcelable(KEY_INSTALLED_ADDON))
 
         doReturn(testContext).`when`(fragment).requireContext()
         val dialog = fragment.onCreateDialog(null)
@@ -131,7 +141,7 @@ class AddonInstallationDialogFragmentTest {
     }
 
     @Test
-    fun `fetching the add-on icon() successfully `() {
+    fun `fetching the add-on icon successfully`() {
         val addon = mock<Addon>()
         val bitmap = mock<Bitmap>()
         val mockedImageView = spy(ImageView(testContext))
@@ -140,13 +150,15 @@ class AddonInstallationDialogFragmentTest {
 
         runBlocking {
             whenever(mockedCollectionProvider.getAddonIconBitmap(addon)).thenReturn(bitmap)
+            assertNull(fragment.arguments?.getParcelable<Bitmap>(KEY_ICON))
             fragment.fetchIcon(addon, mockedImageView, scope).join()
+            assertNotNull(fragment.arguments?.getParcelable<Bitmap>(KEY_ICON))
             verify(mockedImageView).setImageDrawable(Mockito.any())
         }
     }
 
     @Test
-    fun `handle errors while fetching the add-on icon() `() {
+    fun `handle errors while fetching the add-on icon`() {
         val addon = mock<Addon>()
         val mockedImageView = spy(ImageView(testContext))
         val mockedCollectionProvider = mock<AddonCollectionProvider>()
@@ -163,6 +175,32 @@ class AddonInstallationDialogFragmentTest {
                 fail("The exception must be handle in the adapter")
             }
         }
+    }
+
+    @Test
+    fun `allows state loss when committing`() {
+        val addon = mock<Addon>()
+        val mockedCollectionProvider = mock<AddonCollectionProvider>()
+        val fragment = createAddonInstallationDialogFragment(addon, mockedCollectionProvider)
+
+        val fragmentManager = mock<FragmentManager>()
+        val fragmentTransaction = mock<FragmentTransaction>()
+        `when`(fragmentManager.beginTransaction()).thenReturn(fragmentTransaction)
+
+        fragment.show(fragmentManager, "test")
+        verify(fragmentTransaction).commitAllowingStateLoss()
+    }
+
+    @Test
+    fun `cancels icon job on stop`() {
+        val addon = mock<Addon>()
+        val mockedCollectionProvider = mock<AddonCollectionProvider>()
+        val fragment = createAddonInstallationDialogFragment(addon, mockedCollectionProvider)
+
+        val job = mock<Job>()
+        fragment.iconJob = job
+        fragment.onStop()
+        verify(job).cancel()
     }
 
     private fun createAddonInstallationDialogFragment(

--- a/samples/browser/src/main/java/org/mozilla/samples/browser/addons/AddonsFragment.kt
+++ b/samples/browser/src/main/java/org/mozilla/samples/browser/addons/AddonsFragment.kt
@@ -58,8 +58,13 @@ class AddonsFragment : Fragment(), AddonsManagerAdapterDelegate {
             bindRecyclerView(view)
         }
 
-        findPreviousDialogFragment()?.let { dialog ->
-            dialog.onPositiveButtonClicked = onPositiveButtonClicked
+        findPreviousPermissionDialogFragment()?.let { dialog ->
+            dialog.onPositiveButtonClicked = onConfirmPermissionButtonClicked
+        }
+
+        findPreviousInstallationDialogFragment()?.let { dialog ->
+            dialog.onConfirmButtonClicked = onConfirmInstallationButtonClicked
+            dialog.addonCollectionProvider = requireContext().components.addonCollectionProvider
         }
     }
 
@@ -121,11 +126,15 @@ class AddonsFragment : Fragment(), AddonsManagerAdapterDelegate {
     }
 
     private fun isAlreadyADialogCreated(): Boolean {
-        return findPreviousDialogFragment() != null
+        return findPreviousPermissionDialogFragment() != null && findPreviousInstallationDialogFragment() != null
     }
 
-    private fun findPreviousDialogFragment(): PermissionsDialogFragment? {
+    private fun findPreviousPermissionDialogFragment(): PermissionsDialogFragment? {
         return fragmentManager?.findFragmentByTag(PERMISSIONS_DIALOG_FRAGMENT_TAG) as? PermissionsDialogFragment
+    }
+
+    private fun findPreviousInstallationDialogFragment(): AddonInstallationDialogFragment? {
+        return fragmentManager?.findFragmentByTag(INSTALLATION_DIALOG_FRAGMENT_TAG) as? AddonInstallationDialogFragment
     }
 
     private fun showPermissionDialog(addon: Addon) {
@@ -135,7 +144,7 @@ class AddonsFragment : Fragment(), AddonsManagerAdapterDelegate {
 
         val dialog = PermissionsDialogFragment.newInstance(
             addon = addon,
-            onPositiveButtonClicked = onPositiveButtonClicked
+            onPositiveButtonClicked = onConfirmPermissionButtonClicked
         )
 
         if (!isAlreadyADialogCreated() && fragmentManager != null) {
@@ -151,14 +160,7 @@ class AddonsFragment : Fragment(), AddonsManagerAdapterDelegate {
         val dialog = AddonInstallationDialogFragment.newInstance(
             addon = addon,
             addonCollectionProvider = addonCollectionProvider,
-            onConfirmButtonClicked = { _, allowInPrivateBrowsing ->
-                if (allowInPrivateBrowsing) {
-                    requireContext().components.addonManager.setAddonAllowedInPrivateBrowsing(
-                        addon,
-                        allowInPrivateBrowsing
-                    )
-                }
-            }
+            onConfirmButtonClicked = onConfirmInstallationButtonClicked
         )
 
         if (!isAlreadyADialogCreated() && fragmentManager != null) {
@@ -166,17 +168,28 @@ class AddonsFragment : Fragment(), AddonsManagerAdapterDelegate {
         }
     }
 
-    private val onPositiveButtonClicked: ((Addon) -> Unit) = { addon ->
+    private val onConfirmInstallationButtonClicked: ((Addon, Boolean) -> Unit) = { addon, allowInPrivateBrowsing ->
+        if (allowInPrivateBrowsing) {
+            requireContext().components.addonManager.setAddonAllowedInPrivateBrowsing(
+                addon,
+                allowInPrivateBrowsing
+            )
+        }
+    }
+
+    private val onConfirmPermissionButtonClicked: ((Addon) -> Unit) = { addon ->
         addonProgressOverlay.visibility = View.VISIBLE
         isInstallationInProgress = true
 
         val installOperation = requireContext().components.addonManager.installAddon(
             addon,
-            onSuccess = {
-                adapter?.updateAddon(it)
-                addonProgressOverlay.visibility = View.GONE
-                isInstallationInProgress = false
-                showInstallationDialog(it)
+            onSuccess = { installedAddon ->
+                context?.let {
+                    adapter?.updateAddon(installedAddon)
+                    addonProgressOverlay.visibility = View.GONE
+                    isInstallationInProgress = false
+                    showInstallationDialog(installedAddon)
+                }
             },
             onError = { _, e ->
                 // No need to display an error message if installation was cancelled by the user.

